### PR TITLE
JSO Compliance: R2 

### PR DESCRIPTION
**Pull request by MaxenceGollier**
Rewrote `R2_alg.jl` to be more in phase with other JSO solvers. 

- Added a function `SolverCore.solve!( R2Solver, RegularizedNLPModel, stats )` which implements the algorithm.
- Callbacks are now available
- Changed implementations  of the previous R2 functions to fit this new approach but made sure to keep the previous signatures and returns so that nothing else in this package is impacted. This makes the code messy in some places and should be removed. (for example, there is no need for `Fhist`, `Hhist`, etc. in my opinion). Also, the documentation i added does not show these functions.
- On a small benchmark, there seems to be less allocations with this code.

I initially just wanted to add callbacks so i don't know if this is relevant. What do you think @dpo, @geoffroyleconte, @MohamedLaghdafHABIBOULLAH ?

### CODE REVIEW

**Comment by dpo** [comment]
Thank you, @MaxenceGollier! We'll look in detail.

Yes, callbacks are necessary. See https://github.com/JuliaSmoothOptimizers/RegularizedOptimization.jl/pull/131.

File: src/R2_alg.jl [hunk 1]
```diff
@@ -74,60 +125,83 @@ About each iterate xₖ, a step sₖ is computed as a solution of
...
+- `max_time::Float64 = 30.0`: maximum time limit in seconds;
+- `max_iter::Int = 10000`: maximum number of iterations;
+- `verbose::Int = 0`: if > 0, display iteration details every `verbose` iteration;
+- `σmin::T = eps(T)`: minimum value of the regularization parameter;
+- `η1::T = √√eps(T)`: very successful iteration threshold;
+- `η2::T = T(0.9)`: successful iteration threshold;
+- `ν::T = eps(T)^(1 / 5)`: multiplicative inverse of the regularization parameter: ν = 1/σ    ;
+- `γ::T = T(3)`: regularization parameter multiplier, σ := σ/γ when the iteration is very successful and σ := σγ when the iteration is unsuccessful.
+
+The algorithm stops when ``√(ξₖ/νₖ) < atol + rtol*√(ξ₀/ν₀) `` where ξₖ := f(xₖ) + h(xₖ) - φ(sₖ; xₖ) - ψ(sₖ; xₖ)
```
Comment (dpo) [inline src/R2_alg.jl:155]: or ξₖ < 0 but √(-ξₖ/νₖ) < neg_tol.

Also, there is no double backtick in Markdown. Simply use one backtick.

**Review by dpo — CHANGES_REQUESTED** [review]
Just a few quick simple comments for now.

File: src/R2_alg.jl (hunk 1 above)
Comment (MaxenceGollier) [reply src/R2_alg.jl:155]: My bad, I took the double batcktick from `Percival.jl`, it should be removed there too then.

By the way, most of this PR is inspired by the Percival repo.

File: src/R2_alg.jl [hunk 2]
```diff
@@ -74,60 +125,83 @@ About each iterate xₖ, a step sₖ is computed as a solution of
...
+- `max_time::Float64 = 30.0`: maximum time limit in seconds;
+- `max_iter::Int = 10000`: maximum number of iterations;
+- `verbose::Int = 0`: if > 0, display iteration details every `verbose` iteration;
+- `σmin::T = eps(T)`: minimum value of the regularization parameter;
+- `η1::T = √√eps(T)`: very successful iteration threshold;
+- `η2::T = T(0.9)`: successful iteration threshold;
+- `ν::T = eps(T)^(1 / 5)`: multiplicative inverse of the regularization parameter: ν = 1/σ;
+- `γ::T = T(3)`: regularization parameter multiplier, σ := σ/γ when the iteration is very successful and σ := σγ when the iteration is unsuccessful.
+
+The algorithm stops either when `√(ξₖ/νₖ) < atol + rtol*√(ξ₀/ν₀) ` or `ξₖ < 0` and `√(-ξₖ/νₖ) < neg_tol` where ξₖ := f(xₖ) + h(xₖ) - φ(sₖ; xₖ) - ψ(sₖ; xₖ).
```
Comment (dpo) [inline src/R2_alg.jl:155]: To a user who isn’t familiar with the internals of the algorithm, it is not clear what φ and ψ are. Either we could add “… where φ and ψ are adequate models of f and h, respectively”, or we make it simpler with something like “… where √(ξₖ/νₖ) is a stationarity measure.”

File: src/R2_alg.jl [hunk 3]
```diff
@@ -74,60 +125,83 @@ About each iterate xₖ, a step sₖ is computed as a solution of
...
+Changing any of the input arguments will affect the subsequent iterations.
+In particular, setting `stats.status = :user` will stop the algorithm.
+All relevant information should be available in `nlp` and `solver`.
+Notably, you can access, and modify, the following:
+- `solver.xk`: current iterate;
+- `solver.∇fk`: current gradient;
+- `stats`: structure holding the output of the algorithm (`GenericExecutionStats`), which contains, among other things:
+  - `stats.iter`: current iteration counter;
+  - `stats.objective`: current objective function value;
+  - `stats.solver_specific[:smooth_obj]`: current smooth part of the objective function
```
Comment (dpo) [inline src/R2_alg.jl:172]:
```suggestion
  - `stats.solver_specific[:smooth_obj]`: current value of the smooth part of the objective function
```

File: src/R2_alg.jl [hunk 4]
```diff
@@ -74,60 +125,83 @@ About each iterate xₖ, a step sₖ is computed as a solution of
...
+  - `stats.solver_specific[:nonsmooth_objoth]`: current nonsmooth part of the objective function
```
Comment (dpo) [inline src/R2_alg.jl:173]:
```suggestion
  - `stats.solver_specific[:nonsmooth_objoth]`: current value of the nonsmooth part of the objective function
```

File: src/R2_alg.jl [hunk 5]
```diff
@@ -56,9 +61,55 @@ function R2Solver(
...
+    Hobj_hist,
+    Complex_hist,
+  )
+end
+
+
 """
-    R2(nlp, h, options)
-    R2(f, ∇f!, h, options, x0)
+    R2(reg_nlp)
```
Comment (dpo) [inline src/R2_alg.jl:112]:
```suggestion
    R2(reg_nlp; kwargs…)
```

File: src/R2_alg.jl [hunk 6]
```diff
@@ -74,60 +125,83 @@ About each iterate xₖ, a step sₖ is computed as a solution of
...
+- `stats`: structure holding the output of the algorithm (`GenericExecutionStats`), which contains, among other things:
+  - `stats.iter`: current iteration counter;
+  - `stats.objective`: current objective function value;
+  - `stats.solver_specific[:smooth_obj]`: current smooth part of the objective function
+  - `stats.solver_specific[:nonsmooth_objoth]`: current nonsmooth part of the objective function
+  - `stats.status`: current status of the algorithm. Should be `:unknown` unless the algorithm has attained a stopping criterion. Changing this to anything will stop the algorithm, but you should use `:user` to properly indicate the intention.
+  - `stats.elapsed_time`: elapsed time in seconds.
 """
-function R2(nlp::AbstractNLPModel, args...; kwargs...)
+
```
Comment (dpo) [inline src/R2_alg.jl:177]: La fonction `R2(reg_nlp; kwargs…)` (qui se trouve plus bas) doit se trouver juste en-dessous de cette docstring (c’est sa documentation).

File: src/R2_alg.jl [hunk 7]
```diff
@@ -167,53 +253,94 @@ function R2(
...
+  reset!(stats)
+
+  # Retrieve workspace
+  selected = reg_nlp.selected
+  h = reg_nlp.h
+  nlp = reg_nlp.model
+  
+  # Make sure ψ is well shifted (on nlp.meta.x0 in R2Solver Constructor but might be different)
+  if solver.xk != x
```
Comment (dpo) [inline src/R2_alg.jl:337]: If you want to check that the vectors are the same (in the sense that the components are the same), use
```julia
if !all(solver.xk .== x)
```

**Review by dpo — COMMENTED** [review]
Thank you. Here are a few more comments.

I would like to see a comparison of the old and new R2 on the problems we have to make sure the results are the same (number of iterations and final values (f, h, ξ, …).

**Comment by MaxenceGollier** [comment]
Then, for benchmarks,

For this version:
```
julia> using LinearAlgebra, NLPModels, NLPModels, NLPModelsModifiers, RegularizedProblems, ProximalOperators, Random, BenchmarkTools
julia> Random.seed!(0)
julia> bpdn = bpdn_model(1)[1]
julia> λ = norm(grad(bpdn, zeros(bpdn.meta.nvar)), Inf) / 10
julia> h = NormL1(λ)

julia> import Pkg; Pkg.add(url = "https://github.com/MaxenceGollier/RegularizedOptimization.jl.git", rev = "RegularizedModels")
julia> using RegularizedOptimization
julia> options = ROSolverOptions(ν = 1.0, β = 1e16, ϵa = 1e-6, ϵr = 1e-6, verbose = 0)

julia>  @benchmark R2(bpdn, h, options)
BenchmarkTools.Trial: 6852 samples with 1 evaluation.
 Range (min … max):  345.100 μs … 404.146 ms  ┊ GC (min … max):  0.00% … 99.76%
 Time  (median):     558.700 μs               ┊ GC (median):     0.00%
 Time  (mean ± σ):   720.358 μs ±   4.890 ms  ┊ GC (mean ± σ):  11.14% ±  6.57%

          ▃▅█▇▆▄▃▂▁    ▂▄▃▂▁  ▁▂▂▂▁▂▁▁                          ▂
  ▇▃▁▅▃▅▅▁███████████████████████████████▇▇█▇▇▇▇▆▅▆▇▆▇▇▆▅▅▅▃▅▁▅ █
  345 μs        Histogram: log(frequency) by time       1.49 ms <

 Memory estimate: 407.41 KiB, allocs estimate: 7756.
```

and for the old version:
```
julia> using LinearAlgebra, NLPModels, NLPModels, NLPModelsModifiers, RegularizedProblems, ProximalOperators, Random, BenchmarkTools
julia> Random.seed!(0)
julia> bpdn = bpdn_model(1)[1]
julia> λ = norm(grad(bpdn, zeros(bpdn.meta.nvar)), Inf) / 10
julia> h = NormL1(λ)

julia> import Pkg; Pkg.add(url = "https://github.com/JuliaSmoothOptimizers/RegularizedOptimization.jl.git")
julia> using RegularizedOptimization
julia> options = ROSolverOptions(ν = 1.0, β = 1e16, ϵa = 1e-6, ϵr = 1e-6, verbose = 0)

julia> @benchmark R2(bpdn, h, options)
BenchmarkTools.Trial: 276 samples with 1 evaluation.
 Range (min … max):  13.377 ms … 457.888 ms  ┊ GC (min … max): 0.00% … 96.51%
 Time  (median):     15.351 ms               ┊ GC (median):    0.00%
 Time  (mean ± σ):   18.084 ms ±  26.749 ms  ┊ GC (mean ± σ):  9.85% ±  7.25%

   █▆▆▄▁▂▂ ▁      ▃ ▂▂ ▂▄▄▃
  ▆█████████▆▆▅▆▆███████████▁▅▆▅▆▅▁▅▆▅▅▁▆▅▆▁▅▆▁▁▁▁▁▁▅▁▁▁▁▁▁▅▆▅ ▆
  13.4 ms       Histogram: log(frequency) by time      27.2 ms <

 Memory estimate: 3.68 MiB, allocs estimate: 222912.
```

File: src/R2_alg.jl (hunk 2 above)
Comment (MaxenceGollier) [reply src/R2_alg.jl:155]: $\varphi$ and $\psi$ are defined above though. Maybe we can add where $\xi_k$ is the optimal decrease: $\xi_k := ...$

**Comment by MaxenceGollier** [comment]
Changing the line `@benchmark R2(bpdn, h, options)` to `@benchmark R2($bpdn, $h, $options)` don't change anything.

**Comment by dpo** [comment]
@MaxenceGollier I fixed the demos. Could you please rebase your branch? That way, you'll have the latest fixes (builds on FreeBSD + the demos working again). The procedure goes like this:
```bash
> git checkout master
```

If `origin` in your clone points to this repository (you can find out with `git remote -v`), do
```bash
> git pull origin  master
```
If origin points to your fork instead, first do
```bash
> git remote add upstream https://github.com/JuliaSmoothOptimizers/RegularizedOptimization.jl.git
> git pull upstream master
```
Now head back to your branch and rebase
```bash
> git checkout RegularizedModels
> git rebase master
```
Hopefully, there are no conflicts to fix.

**Comment by MaxenceGollier** [comment]
There are still rests of the older version which are just kept to not destroy other implementations in this package that rely on the older version. 
For example, I think the constructor
```
function R2Solver(
  x0::S,
  options::ROSolverOptions,
  l_bound::S,
  u_bound::S;
  ψ = nothing
)
```
is useless. Also, I don't see why we should use `ROSolverOptions` instead of keyword arguments. 
Finally, storing 
```
Fobj_hist::Vector{R}
Hobj_hist::Vector{R}
Complex_hist::Vector{Int}
```
in `R2Solver` is expensive and useless. Should we just keep it this way for the moment and maybe try to remove it later ?

**Comment by dpo** [comment]
Your branch still has conflicts.

Let’s keep “old” features for now. We’ll remove them later. I opened a PR a while ago to remove Fhist, etc., We can update it after this one’s been merged.

**Comment by MaxenceGollier** [comment]
Tests should pass now with 754feea

File: src/R2_alg.jl [hunk 8]
```diff
@@ -74,60 +125,82 @@ About each iterate xₖ, a step sₖ is computed as a solution of
...
+- `max_time::Float64 = 30.0`: maximum time limit in seconds;
+- `max_iter::Int = 10000`: maximum number of iterations;
+- `verbose::Int = 0`: if > 0, display iteration details every `verbose` iteration;
+- `σmin::T = eps(T)`: minimum value of the regularization parameter;
+- `η1::T = √√eps(T)`: very successful iteration threshold;
+- `η2::T = T(0.9)`: successful iteration threshold;
+- `ν::T = eps(T)^(1 / 5)`: multiplicative inverse of the regularization parameter: ν = 1/σ;
+- `γ::T = T(3)`: regularization parameter multiplier, σ := σ/γ when the iteration is very successful and σ := σγ when the iteration is unsuccessful.
+
+The algorithm stops either when `√(ξₖ/νₖ) < atol + rtol*√(ξ₀/ν₀) ` or `ξₖ < 0` and `√(-ξₖ/νₖ) < neg_tol` where ξₖ := f(xₖ) + h(xₖ) - φ(sₖ; xₖ) - ψ(sₖ; xₖ).
```
Comment (dpo) [inline src/R2_alg.jl:155]:
```suggestion
The algorithm stops either when `√(ξₖ/νₖ) < atol + rtol*√(ξ₀/ν₀) ` or `ξₖ < 0` and `√(-ξₖ/νₖ) < neg_tol` where ξₖ := f(xₖ) + h(xₖ) - φ(sₖ; xₖ) - ψ(sₖ; xₖ), and √(ξₖ/νₖ) is stationarity measure.
```

File: src/R2_alg.jl [hunk 9]
```diff
@@ -167,53 +252,97 @@ function R2(
...
+  ) where {T, V}
+
+  reset!(stats)
+
+  # Retrieve workspace
+  selected = reg_nlp.selected
+  h = reg_nlp.h
+  nlp = reg_nlp.model
+  
+  # Make sure ψ is well shifted (on nlp.meta.x0 in R2Solver Constructor but might be different)
```
Comment (dpo) [inline src/R2_alg.jl:338]:
```suggestion
  # Make sure ψ has the correct shift (on nlp.meta.x0 in R2Solver Constructor but might be different)
```

**Review by dpo — COMMENTED** [review]
Just a couple of minor comments.

File: src/R2_alg.jl [hunk 10]
```diff
@@ -74,60 +125,82 @@ About each iterate xₖ, a step sₖ is computed as a solution of
...
-* `f` a function such that `f(x)` returns the value of f at x
-* `∇f!` a function to evaluate the gradient in place, i.e., such that `∇f!(g, x)` store ∇f(x) in `g`.
-
-### Return values
-
-* `xk`: the final iterate
-* `Fobj_hist`: an array with the history of values of the smooth objective
-* `Hobj_hist`: an array with the history of values of the nonsmooth objective
-* `Complex_hist`: an array with the history of number of inner iterations.
+### Arguments
```
Comment (dpo) [inline src/R2_alg.jl:137]:
```suggestion
# Arguments
```

**Review by dpo — CHANGES_REQUESTED** [review]

**Comment by dpo** [comment]
There are lots of large differences in the demos, in particular on the plots showing the number of inner iterations. I wonder if other solvers are able to call the new R2 and if the number of iterations is reported properly. Could you please check?

Among others, these tests show differences:

* bpdn-inner-outer-constr-lm-r2-l0-linf
* bpdn-inner-outer-constr-lm-r2-l1-linf
* bpdn-inner-outer-constr-lmtr-r2-l0-linf
* bpdn-inner-outer-constr-lmtr-r2-l1-linf
* bpdn-inner-outer-r2-l1-l2 (this is R2 alone; the number of iterations goes from 20 to 35)
* bpdn-inner-outer-r2-l1-linf (R2 alone)
* and others.

Those problems are generated randomly but we initialize the seed, so they should be the same.

**Comment by MaxenceGollier** [comment]
Yes they do, when i test locally, all inner iterations are reported correctly. Can you show me with respect to what you are comparing ? I just see the demos-result for this branch (and not for the base one ?)

**Comment by dpo** [comment]
I was comparing to the demos in this PR: https://github.com/JuliaSmoothOptimizers/RegularizedOptimization.jl/pull/138
There have been no changes to the solvers since that PR.

**Comment by MaxenceGollier** [comment]
It is weird because my local branch is up to date with this and when I run this locally on `bpdn-inner-outer-constr-lm-r2-l0-linf`, I get
```
┌ Info:  using LM to solve with
└   h = NormL0{Float64}(0.04974454739546478)
[ Info:  outer    inner     f(x)     h(x)     √ξ1      √ξ        ρ       σ     ‖x‖     ‖s‖   ‖Jₖ‖² reg
[ Info:      1        2  2.1e+00  0.0e+00 4.9e-01 3.2e-01  2.8e+00 1.0e+00 0.0e+00 6.0e-01 2.0e+00 ↘
[ Info:      2        7  1.6e+00  2.0e-01 7.5e-01 7.0e-01  1.6e+00 3.3e-01 6.0e-01 1.3e+00 1.3e+00 ↘
[ Info:      3        8  6.0e-01  4.0e-01 5.5e-01 6.0e-01  1.3e+00 1.1e-01 1.8e+00 1.4e+00 1.1e+00 ↘
[ Info:      4        9  4.4e-02  5.0e-01 1.6e-01 1.8e-01  1.1e+00 3.7e-02 2.8e+00 3.9e-01 1.0e+00 ↘
[ Info:      5       16  9.1e-03  5.0e-01 1.7e-02 2.0e-02  1.0e+00 1.2e-02 3.1e+00 4.6e-02 1.0e+00 ↘
[ Info:      6       15  8.7e-03  5.0e-01 5.9e-04 7.1e-04  1.0e+00 4.1e-03 3.1e+00 1.7e-03 1.0e+00 ↘
[ Info:      7        5  8.7e-03  5.0e-01 8.1e-06 9.7e-06  1.0e+00 1.4e-03 3.1e+00 2.1e-05 1.0e+00 ↘
[ Info:      8        1  8.7e-03  5.0e-01 1.1e-06 1.1e-06          4.6e-04 3.1e+00 1.1e-06 1.0e+00
[ Info: LM: terminating with √ξ1 = 1.069826582248765e-6
```
So I should have the same results as on #138, I don't really see where these plots are generated ? The `stats.solver_specific[:SubsolverCounter]` contains the correct values as well...

**Comment by dpo** [comment]
They’re generated by the scripts in the examples folder.

**Comment by MaxenceGollier** [comment]
The issue should be solved with f0a273cb1784e6832ba67dba14eda966922a9301 now.

File: src/R2_alg.jl [hunk 11]
```diff
@@ -72,62 +123,84 @@ About each iterate xₖ, a step sₖ is computed as a solution of
...
-* `f` a function such that `f(x)` returns the value of f at x
-* `∇f!` a function to evaluate the gradient in place, i.e., such that `∇f!(g, x)` store ∇f(x) in `g`.
-
-### Return values
-
-* `xk`: the final iterate
-* `Fobj_hist`: an array with the history of values of the smooth objective
-* `Hobj_hist`: an array with the history of values of the nonsmooth objective
-* `Complex_hist`: an array with the history of number of inner iterations.
+ψ(s; xₖ) is either h(xₖ + s) or an approximation of h(xₖ + s), ‖⋅‖ is a user-defined norm and σₖ > 0 is the regularization parameter.
```
Comment (MaxenceGollier) [inline src/R2_alg.jl:126]: Is $\|\|\cdot\|\|$ really a user-defined norm ? Isn't it always the $\ell_2$ norm ?
```suggestion
ψ(s; xₖ) is either h(xₖ + s) or an approximation of h(xₖ + s), ‖⋅‖ is a user-defined norm and σₖ > 0 is the regularization parameter.
```

Comment (dpo) [reply src/R2_alg.jl:126]: Oui en effet, c'est la norme 2.

**Pull request merged**